// iamport-react-native/lib/commonjs/utils/IamportUrl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _queryString = _interopRequireDefault(require("query-string"));

var _reactNative = require("react-native");

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class IamportUrl {
  // Android only
  constructor(url) {
    _defineProperty(this, "url", void 0);

    _defineProperty(this, "scheme", void 0);

    _defineProperty(this, "path", void 0);

    _defineProperty(this, "package", void 0);

    this.url = url;
    let splittedUrl = url.replace('://', ' ').split(' ');
    this.scheme = splittedUrl[0];

    if (_reactNative.Platform.OS === 'ios') {
      this.path = this.scheme === 'itmss' ? `https://${splittedUrl[1]}` : this.url;
    } else if (_reactNative.Platform.OS === 'android') {
      if (this.isAppUrl()) {
        if (this.scheme.includes('intent')) {
          let intentUrl = splittedUrl[1].split('#Intent;');
          let host = intentUrl[0];
          let args = intentUrl[1].split(';');

          if (this.scheme !== 'intent') {
            this.scheme = this.scheme.split(':')[1];
            this.path = this.scheme + '://' + host;
          }

          args.forEach(s => {
            if (s.startsWith('scheme')) {
              let scheme = s.split('=')[1];
              this.path = scheme + '://' + host;
              this.scheme = scheme;
            } else if (s.startsWith('package')) {
              let p = s.split('=')[1];
              this.package = p;
            }
          });
        } else {
          this.path = this.url;
        }
      } else {
        this.path = this.url;
      }
    }
  }

  getUrl() {
    return this.url;
  }

  isPaymentOver() {
    return this.url.includes(_constants.IMPConst.M_REDIRECT_URL);
  }

  isAppUrl() {
    return !['http', 'https', 'about:blank'].includes(this.scheme);
  }

  isIframeLoaded() {
    return this.url !== 'about:blank' && !this.url.startsWith(_constants.IMPConst.IMP_SDK_URL);
  }

  getAppUrl() {
    return this.path;
  }

  getMarketUrl() {
    if (_reactNative.Platform.OS === 'ios') {
      switch (this.scheme) {
        case 'kftc-bankpay':
          // 뱅크페이
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id398456030';

        case 'ispmobile':
          // ISP/페이북
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id369125087';

        case 'hdcardappcardansimclick':
          // 현대카드 앱카드
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id702653088';

        case 'shinhan-sr-ansimclick':
          // 신한 앱카드
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id572462317';

        case 'kb-acp':
          // KB국민 앱카드
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id695436326';

        case 'mpocket.online.ansimclick':
          // 삼성앱카드
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id535125356';

        case 'lottesmartpay':
          // 롯데 모바일결제
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id668497947';

        case 'lotteappcard':
          // 롯데 앱카드
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id688047200';

        case 'cloudpay':
          // 하나1Q페이(앱카드)
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id847268987';

        case 'citimobileapp':
          // 시티은행 앱카드
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id1179759666';

        case 'payco':
          // 페이코
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id924292102';

        case 'kakaotalk':
          // 카카오톡
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id362057947';

        case 'lpayapp':
          // 롯데 L.pay
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id1036098908';

        case 'wooripay':
          // 우리페이
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id1201113419';

        case 'com.wooricard.wcard':
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id1499598869';

        case 'nhallonepayansimclick':
          // NH농협카드 올원페이(앱카드)
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id1177889176';

        case 'hanawalletmembers':
          // 하나카드(하나멤버스 월렛)
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id1038288833';

        case 'shinsegaeeasypayment':
          // 신세계 SSGPAY
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id666237916';

        case 'naversearchthirdlogin':
          // 네이버페이 앱 로그인
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id393499958';

        case 'lguthepay-xpay':
          // 페이나우
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id760098906';

        case 'lmslpay':
          // 롯데 L.POINT
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id473250588';

        case 'liivbank':
          // Liiv 국민
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id1126232922';

        case 'supertoss':
          // 토스
          return _constants.IMPConst.IOS_MARKET_PREFIX + 'id839333328';

        default:
          return this.url;
      }
    } else if (_reactNative.Platform.OS === 'android') {
      if (this.package != null) {
        return _constants.IMPConst.ANDROID_MARKET_PREFIX + this.package;
      }

      switch (this.scheme) {
        case _constants.IMPConst.ANDROID_APPSCHEME.ISP:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_ISP;

        case _constants.IMPConst.ANDROID_APPSCHEME.BANKPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_BANKPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.KB_BANKPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_KB_BANKPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.NH_BANKPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_NH_BANKPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.MG_BANKPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_MG_BANKPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.KN_BANKPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_KN_BANKPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.KAKAOPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_KAKAOPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.SMILEPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_SMILEPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.CHAIPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_CHAIPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.PAYCO:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_PAYCO;

        case _constants.IMPConst.ANDROID_APPSCHEME.HYUNDAICARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_HYUNDAICARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.TOSS:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_TOSS;

        case _constants.IMPConst.ANDROID_APPSCHEME.SHINHANCARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_SHINHANCARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.HANACARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_HANACARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.SAMSUNGCARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_SAMSUNGCARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.KBCARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_KBCARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.NHCARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_NHCARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.CITICARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_CITICARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.LOTTECARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_LOTTECARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.LPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_LPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.SSGPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_SSGPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.KPAY:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_KPAY;

        case _constants.IMPConst.ANDROID_APPSCHEME.PAYNOW:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_PAYNOW;

        case _constants.IMPConst.ANDROID_APPSCHEME.WOORIWONCARD:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_WOORIWONCARD;

        case _constants.IMPConst.ANDROID_APPSCHEME.LPOINT:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_LPOINT;

        case _constants.IMPConst.ANDROID_APPSCHEME.KTFAUTH:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_KTFAUTH;

        case _constants.IMPConst.ANDROID_APPSCHEME.LGTAUTH:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_LGTAUTH;

        case _constants.IMPConst.ANDROID_APPSCHEME.SKTAUTH:
          return _constants.IMPConst.ANDROID_MARKET_PREFIX + _constants.IMPConst.ANDROID_PACKAGE.PACKAGE_SKTAUTH;

        default:
          return this.url;
      }
    } else {
      return this.url;
    }
  }

  getQuery() {
    return _queryString.default.parse(this.getStringifiedQuery());
  }

  getStringifiedQuery() {
    const decodedUrl = decodeURIComponent(this.url);
    return _queryString.default.extract(decodedUrl);
  }

  getInicisTransQuery() {
    const {
      m_redirect_url,
      imp_uid,
      merchant_uid
    } = this.getQuery();
    const inicisTransQuery = {
      imp_uid: imp_uid,
      merchant_uid
    };

    if (m_redirect_url !== null && m_redirect_url !== void 0 && m_redirect_url.includes(_constants.IMPConst.M_REDIRECT_URL)) {
      inicisTransQuery.merchant_uid = typeof merchant_uid === 'object' ? merchant_uid[0] : merchant_uid;
    } else {
      inicisTransQuery.merchant_uid = merchant_uid;
    }

    return _queryString.default.stringify(inicisTransQuery);
  }

  async launchApp() {
    if (_reactNative.Platform.OS === 'ios') {
      if (await _reactNative.Linking.canOpenURL(this.url)) {
        return await _reactNative.Linking.openURL(this.getAppUrl());
      }

      try {
        return await _reactNative.Linking.openURL(this.getAppUrl());
      } catch (e) {
        return await _reactNative.Linking.openURL(this.getMarketUrl());
      }
    } else if (_reactNative.Platform.OS === 'android') {
      try {
        return await _reactNative.Linking.openURL(this.getAppUrl());
      } catch (e) {
        return await _reactNative.Linking.openURL(this.getMarketUrl());
      }
    }
  }

}

var _default = IamportUrl;
exports.default = _default;
